### PR TITLE
Add partial typing to NodeNG.infer

### PR DESCRIPTION
**Pull request by DanielNoord**
## Steps

- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Related Issue

### CODE REVIEW

**Review by Pierre-Sassoulas — APPROVED** [review]
Wow, I'm surprised we did not do that earlier, that's a pretty central function 😄 Thanks !

File: astroid/nodes/node_ng.py [hunk 1]
```diff
@@ -124,7 +125,9 @@ def __init__(
         E.g. ClassDef, FunctionDef.
         """
 
-    def infer(self, context=None, **kwargs):
+    def infer(
+        self, context: InferenceContext | None = None, **kwargs
```
Comment (cdce8p) [inline astroid/nodes/node_ng.py:129]:
```suggestion
        self, context: InferenceContext | None = None, **kwargs: Any
```

File: astroid/nodes/node_ng.py [hunk 2]
```diff
@@ -7,11 +7,12 @@
 import pprint
 import sys
 import warnings
-from collections.abc import Iterator
+from collections.abc import Generator, Iterator
 from functools import singledispatch as _singledispatch
 from typing import TYPE_CHECKING, ClassVar, Tuple, Type, TypeVar, Union, cast, overload
```
Comment (cdce8p) [inline astroid/nodes/node_ng.py:12]:
```suggestion
from typing import (
    TYPE_CHECKING,
    Any,
    ClassVar,
    Tuple,
    Type,
    TypeVar,
    Union,
    cast,
    overload,
)
```

**Review by cdce8p — COMMENTED** [review]
Looks good. We should just make sure to type `**kwargs` as well, even if it's only with `Any`.

**Review by cdce8p — APPROVED** [review]

**Pull request merged**